### PR TITLE
ddl: Fix rename partitioned table is not atomic

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/9132

Problem Summary:

In the previous implementation, when running `RENAME` for a partitioned table, it will skip renaming the following partitions if one of the IStorage instance does not exist.

https://github.com/pingcap/tiflash/blob/73929ad45d21e87ec71e7ce0d19e71a4cd5c50bf/dbms/src/TiDB/Schema/SchemaBuilder.cpp#L667-L691

That work well on the versions before v7.2. But after DDL framework refactor since v7.2, there could be a chance that the IStorage instance is not created.
If there are "p1, p2, p3" for a partitioned table, but "p2" is empty. Then the IStorage instance of "p2" is not created. When we need to rename the partitioned table to another database, it will make that "p1" is renamed to the new database, but "p3" remain in the old database. This cause inconsistency.

### What is changed and how it works?

```commit-message

```

1. `applyRenameLogicalTable` will continue the rename for remaining partitions even if it happen to a partition that does not have `IStorage` instance
2. There could be a chance that tiflash restart when applying `applyRenameLogicalTable`, making some partitions is in the old database while some is in the new database. After restart, `syncAllSchema` will call `tryFixPartitionsBelongingDatabase` and check whether there are any partitions that does not belong to the database on disk. If so, rename the partition table to the correct database.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix the issue that tiflash may panic after renaming a partitioned table with empty partition to a new database
```

### CODE REVIEW

File: dbms/src/Debug/MockTiDB.cpp [hunk 1]
```diff
@@ -683,6 +725,45 @@ void MockTiDB::renameTable(const String & database_name, const String & table_na
...
+    RUNTIME_CHECK_MSG(
+        new_database != databases.end(),
+        "new_database is not exist in MockTiDB, new_database_name={}",
+        new_database_name);
+    table->database_id = new_database->second; // set new_database_id
+
+    TableInfo new_table_info = table->table_info;
+    new_table_info.name = new_table_name;
+    auto new_table
+        = std::make_shared<Table>(database_name, table->database_id, new_table_name, std::move(new_table_info));
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockTiDB.cpp:751]:
```suggestion
    auto new_table
        = std::make_shared<Table>(new_database_name, table->database_id, new_table_name, std::move(new_table_info));
```

typo?

File: dbms/src/Debug/MockTiDB.cpp [hunk 2]
```diff
@@ -683,6 +725,45 @@ void MockTiDB::renameTable(const String & database_name, const String & table_na
     version_diff[version] = diff;
 }
 
+void MockTiDB::renameTableTo(
+    const String & database_name,
+    const String & table_name,
+    const String & new_database_name,
+    const String & new_table_name)
+{
+    std::lock_guard lock(tables_mutex);
```
Comment (CalvinNeo) [inline dbms/src/Debug/MockTiDB.cpp:734]: Use scoped_lock

File: dbms/src/Debug/MockTiDB.cpp (hunk 1 above)
Comment (JaySon-Huang) [reply dbms/src/Debug/MockTiDB.cpp]: yes, fixed.

File: dbms/src/Debug/MockTiDB.cpp (hunk 2 above)
Comment (JaySon-Huang) [reply dbms/src/Debug/MockTiDB.cpp]: fixed

**Review by Lloyd-Pottiger — APPROVED** [review]

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp [hunk 3]
```diff
@@ -1439,23 +1453,122 @@ void SchemaBuilder<Getter, NameMapper>::syncAllSchema()
...
+            continue;
+        }
+
+        for (auto tbl_iter = db_ptr->getIterator(context); tbl_iter->isValid(); tbl_iter->next())
+        {
+            const auto table_name = tbl_iter->table()->getTableName();
+            auto opt_tbl_id = SchemaNameMapper::tryGetTableID(table_name);
+            if (!opt_tbl_id)
+            {
+                LOG_WARNING(
```
Comment (CalvinNeo) [inline dbms/src/TiDB/Schema/SchemaBuilder.cpp:1507]: In what case could this happen?

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp [hunk 4]
```diff
@@ -1439,23 +1453,122 @@ void SchemaBuilder<Getter, NameMapper>::syncAllSchema()
...
+
+            LOG_WARNING(
+                log,
+                "FixPartitionsDatabase: try to fix the partition that is not belong to this database, "
+                "db_name={} database_id={} table_id={} new_database_id={}",
+                db_name,
+                database_id,
+                *opt_tbl_id,
+                new_database_id);
+            auto managed_storage = std::dynamic_pointer_cast<IManageableStorage>(tbl_iter->table());
```
Comment (CalvinNeo) [inline dbms/src/TiDB/Schema/SchemaBuilder.cpp:1543]: Also protect the cast?

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp (hunk 3 above)
Comment (JaySon-Huang) [reply dbms/src/TiDB/Schema/SchemaBuilder.cpp:1507]: Cases under "tests/delta-merge-test" may create table with the given table_name rather that "t_${table_id}"

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp (hunk 4 above)
Comment (JaySon-Huang) [reply dbms/src/TiDB/Schema/SchemaBuilder.cpp:1543]: Added

**Pull request merged**